### DISCUSSION TITLE
A simple and stupid question about prometheus_client

### DISCUSSION THREAD

**Discussion by ElapsedSoul** [prometheus/client_golang#938] [Q&A] [answered]
I have a question about prometheus_client.
It's simple protocol for prometheus to pull metrics from exporter.
We can just provide string format by "metric{labels} value" by simple http server.
So what advantages except normalization we get from using prometheus_client library to develop?
Apologize for my bad expression,I just can't reach the points.

backgroud:
I'am new for go and prometheus in one month,but have much programming exp for c++ and python and so on.

**Comment by beorn7** [comment, marked as answer]
You can totally expose Prometheus metrics without using prometheus/client_golang. For very simple cases, that might even be the best. Like you just have one value and want to expose it. Or you have a very specific use case (https://github.com/kubernetes/kube-state-metrics is a famous example of a widely used Go program exposing Prometheus metrics _without_ using prometheus/client_golang .)

prometheus/client_golang solves a whole lot of problems for you, though: Validation, consistency, concurrency-safety, facilitates the actual instrumentation (how to get to the numbers you want to expose), and much more. Just try it out on your own, and you will quickly notice that most cases work much simpler with prometheus/client_golang than without it.

**Reply by ElapsedSoul** [reply]
Thanks a lot.
It seems I'm working on just functional case of metrics exporter while ignored many systematic case which I still don't meet.
